libraries/AP_Button/AP_Button.cpp

```diff
@@ -107,15 +159,91 @@ void AP_Button::update(void)
         hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AP_Button::timer_update, void));        
     }
 
-    if (last_change_time_ms != 0 &&
+    // act on any changes in state
+    {
+        WITH_SEMAPHORE(last_debounced_change_ms_sem);
+        if (last_debounced_change_ms > last_debounce_ms) {
+            last_debounce_ms = last_debounced_change_ms;
+        }
+    }
+
+    // update the PWM state:
+    uint8_t new_pwm_state = 0;
+    for (uint8_t i=0; i<AP_BUTTON_NUM_PINS; i++) {
+        if (!is_pwm_input(i)) {
+            // not a PWM input
+            continue;
+        }
+        const uint16_t pwm_us = pwm_pin_source[i].get_pwm_us();
+        // these values are the same as used in RC_Channel:
+        new_pwm_state |= pwm_us > 1800 && pwm_us < 2200;
```

Comment:
needs shifting

Reply:
Done, and two pwm inputs tested on CubeBlack.